feat: tls for rpc-srv

adhish20 opened the pull request.

adhish20 commented:
To test:
1. generate the key files by running ./scripts/gen-certs.sh
2. Run server : just run rpc-proxy --rpc-server-cert scripts/client.crt --rpc-server-key scripts/client.key --cloud-api-addr https://qa.glaredb.com --cloud-auth-code test
3. just run local -c glaredb://test:adhish@org.0.0.0.0:6444/abc --domain glaredb.com --ca-cert-path scripts/server.crt

adhish20 commented:
the cert names in the testing are weird because I didn't wanted to break pgsrv tests

tychoish commented on crates/glaredb/src/args/local.rs:
```diff
@@ -57,6 +57,14 @@ pub struct LocalClientOpts {
+
+    /// Path to CA certificate for TLS (required for TLS protocol).
+    #[arg(long)]
+    pub ca_cert_path: Option<String>,
```
this is for the client (I am, admittedly new so "client for what?" might be a thing...  shouldn't/wouldn't things fall back to use system certificates, and using this for specific certificates? (presumably in non-testing environments you'd use letsencrypt or something similar.)

In general having the commands/operations have "tls by default, have a flag to disable it" is a good idea.

adhish20 replied:
updated the help line to mention that this is for rpcsrv proxy TLS

tychoish reviewed (commented):
I'm just a guy reading through things, and I *definitely* lack all context, so take all of these comments with a very large grain of salt. 

> the cert names in the testing are weird because I didn't wanted to break pgsrv tests

do they need to be the same certificate? 

speaking of tests? 😂

adhish20 commented on crates/glaredb/src/args/local.rs:
```diff
@@ -57,6 +57,14 @@ pub struct LocalClientOpts {
+
+    /// Domain name against which to verify the server’s TLS certificate (required for TLS protocol).
+    #[arg(long)]
+    pub domain: Option<String>,
```
this is done by tonic by default

adhish20 replied:
this is only client side verification (which is required by tonic). We are only doing this to encrypt the connection.

greyscaled commented:
@tychoish some additional context here: https://github.com/GlareDB/cloud/issues/2164

adhish20 commented:
> I'm just a guy reading through things, and I _definitely_ lack all context, so take all of these comments with a very large grain of salt.
> 
> > the cert names in the testing are weird because I didn't wanted to break pgsrv tests
> 
> do they need to be the same certificate?
> 
> speaking of tests? 😂

we can have separate cert generators for them for the actual tests, these commands are to test that tis is working fine locally

greyscaled commented on scripts/gen-certs.sh:
```diff
@@ -10,3 +10,7 @@ openssl req -new -x509 -days 365 -nodes -text \
         -keyout server.key \
         -subj "/CN=glaredb.com"
+
+openssl req -noenc -newkey rsa -keyout client.key -out client.csr -subj '/CN=glaredb.com' -addext subjectAltName=DNS:glaredb.com
+
+openssl x509 -req -in client.csr -CA server.crt -CAkey server.key -days 365 -out client.crt -copy_extensions copy
```
**Drive by here**:

Would it make sense to gen these in memory in the tests using weak params with the available interface https://docs.rs/openssl/latest/openssl/pkey/index.html and https://docs.rs/openssl/latest/openssl/x509/index.html ?

For example:

```
openssl genpkey  -algorithm RSA -pkeyopt rsa_keygen_bits:512
```

greyscaled replied:
I think we can just remove this after the first pass, as it helped to get things going, but isn't apart of any broader test.

Furthermore `gen-certs` gets called as part of building our image:

https://github.com/GlareDB/glaredb/blob/9e0e4b3dd8b196d151c5515cb58fe91273fa4f91/.github/workflows/image.yaml#L43-L44

https://github.com/GlareDB/glaredb/blob/9e0e4b3dd8b196d151c5515cb58fe91273fa4f91/Dockerfile#L12-L13

and I think having cert files in `scripts/` is confusing (even if gitignored)

universalmind303 commented on crates/glaredb/src/args/mod.rs:
```diff
@@ -49,6 +49,14 @@ pub struct RpcProxyArgs {
+
+    /// Path to server cert to use for TLS (required for TLS protocol).
+    #[clap(long)]
+    pub rpc_server_cert: Option<String>,
+
+    /// Path to server key to use for TLS (required for TLS protocol).
+    #[clap(long)]
+    pub rpc_server_key: Option<String>,
```
could we have aliases for these, 
```sh
./glaredb proxy \
  --rpc-server-cert="path/to/cert" \
  --rpc-server-key="path/to/key"
 ``` 
 
 seems pretty long to me. 
 
 `--cert="cert"` and `--key="key"` seems pretty straightforward to me.

universalmind303 commented on crates/glaredb/src/args/mod.rs:
```diff
@@ -49,6 +49,14 @@ pub struct RpcProxyArgs {
+
+    /// Path to server cert to use for TLS (required for TLS protocol).
+    #[clap(long)]
+    pub rpc_server_cert: Option<String>,
```
If this is required to be a path, we should use `PathBuf` instead of `String`. 

To me, string implies that they could inline the cert `--rpc-server-cert=$MY_CERT` _(which we could add in a later PR if someone requests the functionality)_

I'm pretty sure clap will parse PathBuf's by default.

tychoish commented:
> we can have separate cert generators for them for the actual tests, these commands are to test that tis is working fine locally

This makes sense, I was more saying that we should have some kind of CI-able test of this rather than relying on local/manual testing. (sorry for the shorthand) 

I realize that this is (mostly) for ourselves, and likely to get used mostly in scripts and config files, [which is why i think I tend toward the longer flag names are a bit better, though I don't think its particularly material], but for service-type things, config files seem a bit easier to manage anyway, (I also feel like its easier to add a new flag than it is to remove an old one, so I'd tend to avoid aliases but...

tychoish commented on crates/glaredb/src/rpc_proxy.rs:
```diff
@@ -28,11 +33,33 @@ impl RpcProxy {
+            (Some(cert), Some(key)) => {
+                let cert = std::fs::read_to_string(cert)?;
+                let key = std::fs::read_to_string(key)?;
+                let identity = Identity::from_pem(cert, key);
```
looks like here's the place where we're pulling information (like domain?) out of the server's certificate, I'm inclined to think that the client certs would work the same way.

adhish20 replied:
Identity just stores the cert and the key. Doesn't provide any other information.

tychoish replied:
https://docs.rs/x509-parser/latest/x509_parser/certificate/struct.X509Certificate.html <- looks like this will pull the info out of the certificate.

adhish20 commented:
@greyscaled @tychoish 
Added a config file path that defaults to "/etc/glaredb.conf" which is supposed to be TOML file.
Also added a param to disable tls on the server side. If this looks good, I will do the same for client.
```
[rpc_tls]
server_cert_path = '/Users/adhish/GlareDB/glaredb/server.pem'
server_key_path = '/Users/adhish/GlareDB/glaredb/server.key'
```

tychoish commented on crates/glaredb/src/args/local.rs:
```diff
@@ -58,13 +58,13 @@ pub struct LocalClientOpts {
-    #[arg(long)]
-    pub domain: Option<String>,
+    #[arg(long, value_parser, default_value = "/etc/glaredb.conf", hide = true)]
+    pub tls_conf_path: PathBuf,
```
I'd call this `conf_path`

tychoish reviewed (commented):
in my ideal world, we'd be able to specify default values for this that are reasonable, and then fall back _either_ to a config file or command-line arguments.

ok, having sort of caused an interlude, I think the config file approach is something we should not do as a rush job, and consider what we want more broadly rather than complicate our interface surface on a ticket here, if that makes sense. How does the following sound:

- we do keep the `disable_tls` option (and hide it by default.)
- we add a (hidden) flag for (only) the path to the public certificate (for the client).
- we add a flag to the server (probably for just the key; also hidden) to the server side.
- using the x509 library we can pull the domain out of the parsed certificate. (if this seems reasonable?) 

I think to make this easier to work out, also: 
- if @greyscaled can get you a certificate (that we won't be using, just for testing, so we can pull the cert off a test pod and then rotate) to test against, 
- you can check using the "real" cert (I think it's reasonable to add a shell script and commit unused certs in a `testdata` folder (or similar) that just runs glaredb client/servers to see if it works by eyeball at least right now.)

(everything else about this seems fine to me 😂) 
- 
-

tychoish commented on crates/rpcsrv/src/proxy.rs:
```diff
@@ -76,11 +77,26 @@ impl<A: ProxyAuthenticator> RpcProxyHandler<A> {
+            )
+            .unwrap();
+            let tls_conf = ClientTlsConfig::new()
+                .ca_certificate(Certificate::from_pem(ca))
```
does this mean we're reading the certificate from the file each time a new request is made?

this feels wrong to me (because we'd just read the same thing again and again) but the truth is that cert-manager can replace/rotate the certificate on a running pod, so maybe this is the correct. Maybe there should be a small ticket or a comment to say "cache this in the service struct, and reread occassionally" (the other model is to use a file system watcher to get notified of the events, but that's worse) 

anyway, I'm just thinking outloud; this seems reasonable and I don't think we should change it now, but a comment might be good.
